qiskit/providers/options.py

```diff
@@ -26,10 +26,122 @@ class Options(types.SimpleNamespace):
     options.
     """
 
+    _fields = {}
+
+    def __init__(self, **kwargs):
+        self._fields = {}
+        self._fields.update(kwargs)
+        self.validator = {}
+
+    def __repr__(self):
+        items = (f"{k}={v!r}" for k, v in self._fields.items())
+        return "{}({})".format(type(self).__name__, ", ".join(items))
+
+    def __eq__(self, other):
+        if isinstance(self, Options) and isinstance(other, Options):
+            return self._fields == other._fields
+        return NotImplemented
+
+    def set_validator(self, field, validator_value):
```

Comment:
Would it be more extensible to have a `Validator` class to be registered with the options and have that recognize which options it must validate?

Reply:
I thought about that but I was concerned about that mapping between options and validators in that case. Mostly because the way options are normally used are to have `backend._default_options()` return something like `Options(foo=5, opt=False)` and then the fields in the options object should be frozen to just `foo` and `opt`. I was worried about the case like:

```
Validator = Validator(foo=(0, 10), opt=bool, choices=['a', 'b', 'c'])
options = Options(foo=5, opt=False, validator=validator)
```
(there's also the potential name conflict with `validator`)

and the user wouldn't know that `choices` wouldn't do anything. Or were you thinking of some other pattern?